Add XDG_DATA_DIRS support for using plugins from multiple locations

waveywaves opened the pull request:
fixes https://github.com/helm/helm/issues/7529

This will add support for XDG_DATA_DIRS env var which will
allow for using pugins from multiple locations defined in XDG_DATA_DIRS

waveywaves commented:
cc @mattfarina 
Need some clarity on a few pointers below 

- We need to Introduce HELM_DATA_DIRS envVar which can leverage XDG_DATA_DIRS for information.

- XDG_DATA_DIRS already exists as an envVar in our PCs, how do we segregate HELM related EnvVars ? OR do we set them directly in HELM_DATA_DIRS if we plan to introduce them ?

- Plugin Download to multiple DATA DIRS can be supported with proper implementation, https://github.com/helm/helm/issues/7528. Does the global installation of plugin need to be facilitated by a `--global` | `-g` flag ? As it makes sense to keep plugin install local to the preferred HELM_PLUGINS directory.

technosophos commented:
What is the status on this PR?

jdolitsky commented:
@waveywaves can you check out the CI failure?
```
internal/test/ensure/ensure.go:34: File is not `gofmt`-ed with `-s` (gofmt)
	os.Setenv(xdg.DataDirsEnvVar,base+";")
cmd/helm/root_test.go:34: File is not `gofmt`-ed with `-s` (gofmt)
		envvars                                     map[string]string
```

waveywaves commented:
@technosophos Can you PTAL ?

technosophos commented:
This one still needs a rebase, but after that I'll work with @adamreese on us getting it reviewed.

waveywaves commented:
@technosophos done with rebase. Do check it oit

hickeyma reviewed (commented):
I wonder should the PR possibly look to support general case of an overall override, similar to a HELM_HOME in v2? Issue #7919 could be supported by the pr also.

technosophos commented:
Please hold on this PR until we get clarity on #7529. I don't think this PR solves the issue, but we need to make sure we understand the issue.

technosophos commented on pkg/helmpath/lazypath.go:
```diff
@@ -47,3 +48,13 @@ func (l lazypath) configPath(elem ...string) string {
 }
+
+// dataPath defines the base directory relative to which user specific data files should be stored.
+func (l lazypath) dataDirsPath(elem ...string) string {
```
The spec says that local data dir (XDG_LOCAL_DATA) should be read first, and should be pre-pended to the list of XDG_DATA_DIRS.

technosophos replied:
Sorry... `XDG_DATA_HOME`, not `XDG_LOCAL_DATA`. See https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

technosophos commented on pkg/helmpath/lazypath_unix.go:
```diff
@@ -43,3 +44,10 @@ func configHome() string {
+// dataDirs defines the directories relevant directories in use separated with a colon
+//
+// $XDG_DATA_DIRS is usually set by the system
+func dataDirs() string {
```
This does not seem to be what the spec suggests doing.

The spec seems to suggest that the system can provide additional paths on XDG_DATA_DIRS that application code (like Helm) may/should consider as other locations where data may be stored. But cache and config are not also to be considered data dirs. This would imply that stuff in the cache was to be treated as data dirs (and likewise with config). I don't think that is what the spec intends.

technosophos reviewed (changes requested):
I think a few different things got put in the same PR.

- the XDG spec provides a convention for naming three different _kinds_ of directories: Cache, Data, and Config. Different data is stored in each of these. They are not interchangeable. So we should never return, say, the config directory when a request is made for a data directory.
- The XDG_DATA_DIRS directory is supplied by the system, and is to be _appended_ to the user-supplied XDG_DATA_HOME directory. Helm should never override or presume to generate a value for XDG_DATA_DIRS. Think of it as communication from the system to the user and to Helm.
- What @mattfarina is asking for in #7529 is that when Helm loads plugins, it also checks the XDG_DATA_DIRS for additional plugins, the way `os/exec.LookPath` looks at each of the subdirs in `PATH` to find the first matching executable.

This seems to have support for #7919. There is some debate over whether we ought to do that. So maybe we should pick one of the two issues and solve it, rather than have both issues in the same PR.

waveywaves commented:
@technosophos I understand this better now. Let me update the PR accordingly.

waveywaves commented:
Continuing work on this today and finishing off.

waveywaves committed:
Add XDG_DATA_DIRS support for using plugins from multiple locs

This allows the user to set their plugin directories in
XDG_DATA_DIRS and based on the precedence, all the locations will
be searched for the particular plugin and be used.

Signed-off-by: Vibhav Bobade <vibhav.bobde@gmail.com>

waveywaves commented:
@technosophos @adamreese can you PTAL ? I have updated to search for plugins in the XDG_DATA_DIRS dirs

waveywaves commented:
@technosophos @adamreese can you PTAL ? Have updated it based on your comments.
